### PR TITLE
HIVE-26243: Add vectorized implementation of the 'ds_kll_sketch' UDAF

### PR DESCRIPTION
**Pull request by asolimando**
### What changes were proposed in this pull request?

We add a vectorized implementation for the `ds_kll_sketch` UDAF

### Why are the changes needed?

When this UDAF is used either alone or at the side of other vectorizable functions, it will benefit from a performance speed-up.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

```mvn test -Dtest=TestMiniLlapLocalCliDriver -Dqfile="compute_kll_sketch.q" -Dtest.output.overwrite -pl itests/qtest -Pitests```

and 

```
mvn test -Dtest=TestMiniLlapLocalCliDriver -Dqfile="sketches_rewrite_rank_partition_by.q,sketches_rewrite_rank.q,sketches_rewrite_percentile_disc.q,sketches_rewrite_ntile_partition_by.q,sketches_rewrite_ntile.q,sketches_rewrite_cume_dist_partition_by.q,sketches_rewrite_cume_dist.q,sketches_materialized_view_rank.q,sketches_materialized_view_percentile_disc.q,sketches_materialized_view_ntile.q,sketches_materialized_view_cume_dist.q" -Dtest.output.overwrite -pl itests/qtest -Pitests
```

### CODE REVIEW

File: standalone-metastore/pom.xml [hunk 1]
```diff
@@ -181,6 +182,17 @@
         <artifactId>commons-lang3</artifactId>
         <version>${commons-lang3.version}</version>
       </dependency>
+      <dependency>
+        <groupId>org.apache.datasketches</groupId>
```
Comment (kgyrtkirk) [inline standalone-metastore/pom.xml:186]: I think this has nothing to do with this patch

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/kll/KllUtils.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * KLL serialization utilities.
+ */
+public class KllUtils {
```
Comment (kgyrtkirk) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/kll/KllUtils.java:34]: unrelated to changes?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+import org.apache.datasketches.kll.KllFloatsSketch;
+import org.apache.hadoop.hive.common.histogram.kll.KllUtils;
+import org.apache.hadoop.hive.common.type.HiveDecimal;
+import org.apache.hadoop.hive.ql.util.JavaDataModel;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+public class KllHistogramEstimator {
```
Comment (kgyrtkirk) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java:28]: remove these classes from the metastore...

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/kll/KllUtils.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  /**
+   * KLL is serialized according to what provided by data-sketches library
+   * @param out output stream to write to
+   * @param kll KLL sketch that needs to be serialized
+   * @throws IOException if an error occurs during serialization
+   */
+  public static void serializeKll(OutputStream out, KllFloatsSketch kll) throws IOException {
+    out.write(kll.toByteArray());
+  }
```
Comment (kgyrtkirk) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/kll/KllUtils.java:48]: do we really need these methods/classes/etc? 

this really just binds the class name; the method name with no good reason...no interface; what's the architectural value here I don't see?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimatorFactory.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+   */
+  public static KllHistogramEstimator getKllHistogramEstimator(byte[] buf, int start, int len) {
+    return new KllHistogramEstimator(KllUtils.deserializeKll(buf, start, len));
+  }
+
+  /**
+   * This method creates an empty histogram estimator with a KLL sketch with k=200.
+   * @return an empty histogram estimator with a KLL sketch with k=200
+   */
+  public static KllHistogramEstimator getEmptyHistogramEstimator() {
```
Comment (kgyrtkirk) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimatorFactory.java:54]: there are 3 `getEmptyHistogramEstimator` -s from which only 2 is being used; and I think 1 will be too much already...do we need these at all? what's the point of introducing these???

File: ql/src/gen/vectorization/UDAFTemplates/VectorUDAFComputeKLL.txt [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+          }
+        }
+      }
+    }
+#ENDIF COMPLETE
+#IF MERGING
+    if (inputColumn.isRepeating) {
+      if (!inputColumn.isNull[0] && inputColumn.length[0] > 0) {
+        myagg.prepare();
+        KllHistogramEstimator mergingKLL = KllHistogramEstimatorFactory.getKllHistogramEstimator(
```
Comment (kgyrtkirk) [inline ql/src/gen/vectorization/UDAFTemplates/VectorUDAFComputeKLL.txt:124]: do we really have to create and bind 1-by-1 vectorized implementations for the datasketches functions ?
function classes kinda do the same stuff... like `DATA_TO_SKETCH` ;

File: ql/src/gen/vectorization/UDAFTemplates/VectorUDAFComputeKLL.txt [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+#ENDIF COMPLETE
+import org.apache.hadoop.hive.ql.exec.vector.VectorAggregationBufferRow;
+import org.apache.hadoop.hive.ql.exec.vector.VectorAggregationDesc;
+import org.apache.hadoop.hive.ql.exec.vector.VectorizedRowBatch;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFEvaluator;
+import org.apache.hadoop.hive.ql.util.JavaDataModel;
+
+/**
+ * Generated from template VectorUDAFComputeKLL.txt.
```
Comment (kgyrtkirk) [inline ql/src/gen/vectorization/UDAFTemplates/VectorUDAFComputeKLL.txt:37]: this seems to be the vectorized pair of `ds_kll_sketch` or not? 

can we keep the naming conventions? it will be hard to follow what-is-what....without doing so

**Comment by asolimando** [comment]
Before going into the single discussions, the general answer to all the above comments boils down to "I am trying to keep consistency with what was done here for vectorizing HyperLogLog function": https://github.com/apache/hive/pull/1824/files

I sense that you don't like how that PR was designed, but since they are very close in spirit, and that their code is used side by side, I thought it was important to keep them consistent.

If we need to rework the current PR, they won't match anymore, unless we rework the HLL design and implementation too, and this has its own share of cons...

Assuming we go for the refactoring, most of the comments are too sketchy to give appropriate guidance over an alternative design/implementation, I will need to ask you to elaborate more on them.

For instance, you seem to be suggesting to remove all helper classes/methods etc. Since it does not seem feasible to inline all the code now sitting in the helper methods/classes directly in the vectorized implementation, I guess you want to place it someplace else, but I can't really decide based on your comment.

For the couple of currently unused methods, I will need them in a PR depending on this one: https://issues.apache.org/jira/browse/HIVE-26221: I can remove them now and re-introduce them later, if preferable. Once again they mimic HLL methods (both naming and usage, since HLL and KLL methods will be used side by side in most places, it helps reading what's happening, see [LongColumnStatsAggregator.java#L104-L111](https://github.com/asolimando/hive/blob/master-histograms_stats_rebased/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java#L104-L111), for instance).

**Comment by kgyrtkirk** [comment]
PR-1824 has nothing to do with datasketches; I don't know how you followed it's conventions but you might end up in trouble...because DS also has a HLL implementation...wouldn't that conflict with the existing one?

note: PR-1824 named the file `VectorUDAFComputeBitVector.txt` and internally named the method `compute_bit_vector_hll` ; I think the class name should have contained the `Hll` keyword

I think that the file name `VectorUDAFComputeKLL.txt` is not connected at all to the `ds_kll_sketch` function its about to vectorize...and as such its a bit confusing....

The current implementation doesn't really look forward: I think we have 20 *sketch* function from datasketches already exposed as inside Hive which could be vectorized; I think they are behind the same api cover...so just vectorizing the KLL one without any sight forward and taking "ideas" from the old hll codepath doesn't seem the best idea to me...
```
grep ^ds_ ql/src/test/results/clientpositive/llap/show_functions.q.out|grep _sketch$
```
no need to do everything in 1 patch - but this is pretty much just copy-pasting the existing hll txtfile substituted to kll here and there...so we should do that 20 times?

> For instance, you seem to be suggesting to remove all helper classes/methods etc
I don't think those changes neccessary in the *metastore* for a vectorization of this function? 

HIVE-26221 is something which have changes - but has no real end-user accessible value - and as such I don't think its ready.

**Comment by asolimando** [comment]
> PR-1824 has nothing to do with datasketches; I don't know how you followed it's conventions but you might end up in trouble...because DS also has a HLL implementation...wouldn't that conflict with the existing one?

HyperLogLog is one of the few examples of vectorized functions over a "complex" data type, it seemed like a good candidate to me.

> note: PR-1824 named the file `VectorUDAFComputeBitVector.txt` and internally named the method `compute_bit_vector_hll` ; I think the class name should have contained the `Hll` keyword
> 
> I think that the file name `VectorUDAFComputeKLL.txt` is not connected at all to the `ds_kll_sketch` function its about to vectorize...and as such its a bit confusing....

I agree on that, maybe `VectorUDAF_ds_kll_sketch` would be clearer.

> The current implementation doesn't really look forward: I think we have 20 _sketch_ function from datasketches already exposed as inside Hive which could be vectorized; I think they are behind the same api cover...so just vectorizing the KLL one without any sight forward and taking "ideas" from the old hll codepath doesn't seem the best idea to me...
> 
> ```
> grep ^ds_ ql/src/test/results/clientpositive/llap/show_functions.q.out|grep _sketch$
> ```
> 
> no need to do everything in 1 patch - but this is pretty much just copy-pasting the existing hll txtfile substituted to kll here and there...so we should do that 20 times?

Sketches are behind the same api for methods like `update()` and `serialize()`, but they also have differences, like the supported constructors, it seems hard to have a uniform UDAF for all of them. I have omitted the constructor for supporting the K param of Kll sketches in the current PR, but it will be needed later on (I have implemented it as suggested here: https://github.com/apache/hive/pull/1824#issuecomment-756588127), so we will have to have a specialized version of the KLL UDAF anyway, and the same for all the other 9 sketches, I don't see a way out of this.

> > For instance, you seem to be suggesting to remove all helper classes/methods etc
> > I don't think those changes neccessary in the _metastore_ for a vectorization of this function?
> 
> [HIVE-26221](https://issues.apache.org/jira/browse/HIVE-26221) is something which have changes - but has no real end-user accessible value - and as such I don't think its ready.

[HIVE-26221](https://issues.apache.org/jira/browse/HIVE-26221) already sensibly improves range filter estimation for runtime statistics that can be used by Tez, of course exploiting statistics in query planning is the next obvious step, but I am a fan of iterative improvements, especially which such a big patch spanning few modules. Let's have this discussion offline as it is not key here, I just wanted to highlight the needs behind the current patch by linking to this ticket.

I am trying to figure out a better place to put an interface plus the concrete implementation(s) of the "complex" object incapsulating the different sketches: the only examples I could find are `HyperLogLog` (stored in the `metastore-server` module) and `BloomFilter`/`BloomKFilter` (stored in the `storage-api` module). As of now I have followed the `HyperLogLog` example but you seem to disagree, `storage-api` seems pretty arbitrary too, so I am not sure how to proceed here, do you have any suggestions?

**Comment by asolimando** [comment]
Please keep the PR open

**Comment by asolimando** [comment]
Please keep it open

File: standalone-metastore/metastore-server/pom.xml [hunk 8]
```diff
@@ -44,6 +44,16 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.apache.datasketches</groupId>
+      <artifactId>datasketches-hive</artifactId>
+      <exclusions>
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/pom.xml:50]: could we declare the exclusions in mvn dependency management so we won't have to define them every time we use the dependency?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java [hunk 9]
```diff
@@ -4461,17 +4464,21 @@ public static ImmutablePair<VectorAggregationDesc,String> getVectorAggregationDe
...
-    VectorizedUDAFs annotation =
-        AnnotationUtils.getAnnotation(evaluator.getClass(), VectorizedUDAFs.class);
-    if (annotation == null) {
-      String issue =
-          "Evaluator " + evaluator.getClass().getSimpleName() + " does not have a " +
-          "vectorized UDAF annotation (aggregation: \"" + aggregationName + "\"). " +
-          "Vectorization not supported";
-      return new ImmutablePair<VectorAggregationDesc,String>(null, issue);
+    Class<? extends VectorAggregateExpression>[] vecAggrClasses;
+    if (aggregationName.equals("ds_kll_sketch")) {
```
[the remark this answers is not in the document]
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java:4468]: I have drafted a solution using a private enumeration, please let me know if I got what you meant

**Comment by asolimando** [comment]
Rebased on master in order to take advantage of PR-[3637](https://github.com/apache/hive/pull/3637) to support KLL sketch size parameter in the constructor (otherwise we will miss vectorization opportunities when the UDAF is used in its version specifying the sketch size). Commits implementing this will come shortly.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java [hunk 10]
```diff
@@ -304,7 +327,7 @@ public enum EnabledOverride {
         nameMap.put(
             vectorizationEnabledOverride.name().toLowerCase(), vectorizationEnabledOverride);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java:328]: could you please create toString() method that returns name().toLowerCase() and use that

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimatorFactory.java (hunk 5 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimatorFactory.java:54]: i think this is still valid, please check and remove unused factory methods

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimatorFactory.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  private KllHistogramEstimatorFactory() {
+    throw new AssertionError("Suppress default constructor for non instantiation");
+  }
+
+  /**
+   * This function deserializes the serialized KLL histogram estimator from a byte array.
+   * @param buf to deserialize
+   * @return KLL histogram estimator
+   */
+  public static KllHistogramEstimator getKllHistogramEstimator(byte[] buf) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimatorFactory.java:35]: is it used?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java (hunk 3 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java:28]: should we keep those classes in hive-exec module?

File: standalone-metastore/pom.xml (hunk 1 above)
Comment (deniskuzZ) [reply standalone-metastore/pom.xml:184]: why do you need this, it's already defined in root pom DM?

File: ql/src/gen/vectorization/UDAFTemplates/VectorUDAFComputeKLL.txt (hunk 7 above)
Comment (deniskuzZ) [reply ql/src/gen/vectorization/UDAFTemplates/VectorUDAFComputeKLL.txt:37]: I am not familiar with naming conventions here, but I think Zoli is saying that a more appropriate name would be VectorUDAFComputeDsKllSketch

File: standalone-metastore/pom.xml (hunk 1 above)
Comment (asolimando) [reply standalone-metastore/pom.xml:184]: `standalone-metastore` does not inherit from the root pom

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java (hunk 10 above)
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java:328]: I haven't introduced that enum but since we are improving slightly over the existing I have done that.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java [hunk 12]
```diff
@@ -244,7 +247,7 @@
 
 public class Vectorizer implements PhysicalPlanResolver {
 
-  protected static transient final Logger LOG = LoggerFactory.getLogger(Vectorizer.class);
+  protected static final Logger LOG = LoggerFactory.getLogger(Vectorizer.class);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java:250]: why was the transient modifier removed?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java (hunk 3 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java:28]: maybe `org.apache.hadoop.hive.ql.udf.generic.histogram` in hive-exec?  @szlta what do you think?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java (hunk 12 above)
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java:250]: Sonar marked it as [code smell](https://sonarcloud.io/project/issues?issues=AYJpdDZw9k-CsauE5Bp5&open=AYJpdDZw9k-CsauE5Bp5&id=apache_hive) because for non-`Serializable` classes the modifier is a no-op and it's suggested to remove it.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java (hunk 3 above)
Comment (asolimando) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java:28]: I like your suggestion, the only concern I have is that `KllHistogramEstimator` is the histogram's equivalent of [`NumDistinctValueEstimator`](https://github.com/apache/hive/blob/8b48ce50684791d53dcb942958de233263cc421f/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/ndv/NumDistinctValueEstimator.java) for NDV, I thought it might be confusing to place it elsewhere, but I am fine with whatever you and @szlta suggest.

Comment (szlta) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java:28]: `org.apache.hadoop.hive.ql.udf.generic.histogram` sounds good to me

Comment (asolimando) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/common/histogram/KllHistogramEstimator.java]: I have moved the classes to `org.apache.hadoop.hive.ql.udf.generic.histogram` as suggested, it looks indeed better, thanks!

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/histogram/kll/KllUtils.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    out.write(kll.toByteArray());
+  }
+
+  /**
+   * This function deserializes the serialized KLL sketch from a stream.
+   * @param in input stream to be deserialized
+   * @return KLL sketch
+   * @throws IOException if errors occur while reading the stream
+   */
+  public static KllFloatsSketch deserializeKll(InputStream in) throws IOException {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/histogram/kll/KllUtils.java:56]: please make sure you do not have a dead code, haven't found any usage of this method

**Review by deniskuzZ — APPROVED** [review]
LGTM +1, pending tests

**Comment by asolimando** [comment]
Rebasing on master

**Comment by asolimando** [comment]
@deniskuzZ tests are green, ready to be merged when you have a moment, thanks!
